src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -293,53 +357,52 @@ private ClassDesc getNearestClassWithSameName(String className, String superClas
      * Get the difference between class declaration name matching count. If the difference between
      * them is zero, then their depth is compared to obtain the result.
      *
-     * @param superClassName name of the super class
+     * @param superOrOuterClassName name of the super class or the outer class
      * @param firstClass first input class
      * @param secondClass second input class
+     * @param shouldSearchAnonSuperClass should search for the super class of anonymous inner class
      * @return difference between class declaration name matching count
      */
-    private static int getClassDeclarationNameMatchingCountDiff(String superClassName,
-                                                                ClassDesc firstClass,
-                                                                ClassDesc secondClass) {
-        int diff = Integer.compare(
-            CheckUtil
-                .typeDeclarationNameMatchingCount(superClassName, secondClass.getQualifiedName()),
-            CheckUtil
-                .typeDeclarationNameMatchingCount(superClassName, firstClass.getQualifiedName()));
+    private static int getClassDeclarationNameMatchingCountDiff(String superOrOuterClassName,
+        ClassDesc firstClass, ClassDesc secondClass, boolean shouldSearchAnonSuperClass) {
```

Comment:
to avoid this stuff with flag and ugly (as for me) code, I can suggest following
```
private static int getClassDeclarationNameMatchingCountDiff(
        ClassDesc firstClass, ClassDesc secondClass, Function<String, Integer> countProvider){
diff = Integer.compare(
countProvider.apply(secondClass.getQualifiedName()),
countProvider.apply(firstClass.getQualifiedName()),
)
// rest is same
```

So, when you invoke this code, you will have something like
```
getClassDeclarationNameMatchingCountDiff(firstClass, secondClass,
 name -> CheckUtil.typeDeclarationMatchingCountAnonSuperClass(superOrOuterClassName, name)
```

??

Reply:
Noice.
It is a lot better than using flags, this can be used to replace flags in a lot of cases, noted down.